mempool/src/shared_mempool/network.rs

```diff
@@ -103,38 +103,66 @@ impl<NetworkClient: NetworkClientInterface<MempoolSyncMsg>> MempoolNetworkInterf
         }
     }
 
-    /// Add a peer to sync states, and returns `false` if the peer already is in storage
-    pub fn add_peer(&self, peer: PeerNetworkId, metadata: ConnectionMetadata) -> bool {
+    /// Returns peers to add (with metadata) and peers to disable
+    fn determine_peers_to_add_and_disable(
+        &self,
+        updated_peers: &HashMap<PeerNetworkId, PeerMetadata>,
+    ) -> (Vec<(PeerNetworkId, ConnectionMetadata)>, Vec<PeerNetworkId>) {
+        let sync_states = self.sync_states.read();
+        let to_disable: Vec<_> = sync_states
+            .keys()
+            .filter(|previous_peer| !updated_peers.contains_key(previous_peer))
+            .copied()
+            .collect();
+        let to_add: Vec<_> = updated_peers
+            .iter()
+            .filter(|(peer, _)| !sync_states.contains_key(peer))
+            .map(|(peer, metadata)| (*peer, metadata.get_connection_metadata()))
+            .filter(|(peer, metadata)| self.is_upstream_peer(peer, Some(metadata)))
+            .collect();
+        (to_add, to_disable)
+    }
+
+    /// Returns newly added peers
+    fn add_and_disable_peers(
+        &self,
+        to_add: &[(PeerNetworkId, ConnectionMetadata)],
+        to_disable: &[PeerNetworkId],
+    ) {
         let mut sync_states = self.sync_states.write();
-        let is_new_peer = !sync_states.contains_key(&peer);
-        if self.is_upstream_peer(&peer, Some(&metadata)) {
-            // If we have a new peer, let's insert new data, otherwise, let's just update the current state
-            if is_new_peer {
+        for (peer, metadata) in to_add.iter().cloned() {
+            if self.is_upstream_peer(&peer, Some(&metadata)) {
```

Comment:
nit: do we still need this check if we only add upstream peers above?

Reply:
Good catch!